components/snap_recovery/src/leader_keeper.rs

```diff
@@ -0,0 +1,260 @@
+            )
+            .field(
+                "campaign_failed",
+                &format_args!("{}", ellipse(&self.campaign_failed, 8)),
+            )
+            .finish()
+    }
+}
+
+impl<EK, Router> LeaderKeeper<EK, Router>
+where
+    EK: KvEngine,
+    Router: CasualRouter<EK> + SignificantRouter<EK> + 'static,
+{
+    pub fn new(router: Router, to_keep: impl IntoIterator<Item = u64>) -> Self {
+        Self {
+            router,
+
+            not_leader: to_keep.into_iter().collect(),
+            _ek: PhantomData,
+        }
+    }
+
+    pub async fn elect_and_wait_all_ready(&mut self) {
+        loop {
+            let now = Instant::now();
+            let res = self.step().await;
+            info!("finished leader keeper stepping."; "result" => ?res, "take" => ?now.elapsed());
+            GLOBAL_TIMER_HANDLE
+                .delay(now + Duration::from_secs(10))
+                .compat()
+                .await
+                .expect("wrong with global timer, cannot stepping.");
+            if res.failed_leader.is_empty() {
+                return;
+            }
+        }
+    }
+
+    pub async fn step(&mut self) -> StepResult {
+        const CONCURRENCY: usize = 256;
+        let r = Mutex::new(StepResult::default());
+        let success = Mutex::new(HashSet::new());
+        for batch in &self.not_leader.iter().chunks(CONCURRENCY) {
+            let tasks = batch.map(|region_id| async {
+                match self.check_leader(*region_id).await {
+                    Ok(_) => {
+                        success.lock().unwrap().insert(*region_id);
+                        return;
+                    }
+                    Err(err) => r.lock().unwrap().failed_leader.push((*region_id, err)),
+                };
+
+                if let Err(err) = self.force_leader(*region_id) {
+                    r.lock().unwrap().campaign_failed.push((*region_id, err));
+                }
+            });
+            futures::future::join_all(tasks).await;
+        }
+        success.lock().unwrap().iter().for_each(|i| {
+            debug_assert!(self.not_leader.remove(i));
+        });
+        r.into_inner().unwrap()
+    }
+
+    async fn check_leader(&self, region_id: u64) -> Result<()> {
+        let (cb, fut) = paired_future_callback();
+        let msg = SignificantMsg::LeaderCallback(Callback::<EK::Snapshot>::read(cb));
+        self.router.significant_send(region_id, msg)?;
+        let resp = fut
+            .await
+            .map_err(|_err| Error::Other("canceled by store".into()))?;
+        let header = resp.response.get_header();
+        if header.has_error() {
+            return Err(Error::Other(box_err!(
+                "got error: {:?}",
+                header.get_error()
+            )));
+        }
+        Ok(())
+    }
+
+    fn force_leader(&self, region_id: u64) -> Result<()> {
+        let msg = CasualMessage::Campaign;
+        self.router.send(region_id, msg)?;
```

Comment:
Is it necessary to keep the log "region starts to campaign"?

Reply:
I'm not sure, somethings it gets so verbose that some of logs will be dropped.